src/util/focus-trap.ts

```diff
@@ -0,0 +1,142 @@
+import {DOCUMENT} from '@angular/common';
+import {Inject, Injectable, NgZone} from '@angular/core';
+
+const FOCUSABLE_ELEMENTS_SELECTOR = [
+  'a[href]', 'button:not([disabled])', 'input:not([disabled]):not([type="hidden"])', 'select:not([disabled])',
+  'textarea:not([disabled])', '[contenteditable]', '[tabindex]:not([tabindex="-1"])'
+].join(', ');
+
+enum DIRECTION {
+  BACKWARD,
+  FORWARD
+}
+
+/**
+ * Class that enforce the browser focus to be trapped inside a DOM element.
+ *
+ * The implementation is rather simple, the class add a `focusin` listener on the document with capture phase.
+ * Any focus event will then be caught, and therefore the class will only allow the one for elements contained inside
+ * it's own element.
+ *
+ * In case the element is not contained, the class will determine which new element has to be focused based on the `tab`
+ * navigation direction.
+ *
+ * Should not be used directly. Use only via {@link NgbFocusTrapFactory}
+ */
+export class NgbFocusTrap {
+  private _removeDocumentListener;
+  private _direction: DIRECTION = DIRECTION.FORWARD;
+  private _previouslyFocused: HTMLElement | null = null;
+  private _endOfDocument: HTMLElement | null = null;
+
+  /**
+   * Register the focustrap instance to initially put the focus on specific element
+   *
+   * Simply add the `ngbAutofocus` attribute on the given element such as `<button ngbAutofocus></button>`
+   * Any previously focused element outside of the focustrap will be remembered and restored on destroy
+   */
+  set autofocus(value: boolean) {
+    if (value === true && this._previouslyFocused === null) {
+      this._previouslyFocused = document.activeElement as HTMLElement;
+      this._focusInitial();
+    }
+  }
+
+  /**
+   * Guess the next focusable element.
+   * Computation is based on specific CSS selector and [tab] navigation direction
+   */
+  get focusableElement(): HTMLElement {
```

Comment:
It is neither used publicly nor in tests. Why have a public getter for it instead of a private function?

Reply:
I used to have tests in one of the previous version. I will make it private